### PR TITLE
Implementation of a rockcraft-built nginx rock

### PR DESCRIPTION
**Pull request by gruyaume**
# OVERVIEW

This PR implements a rockcraft-built rock for nginx based on Ubuntu 22.04. It also takes care of publishing it to github's package registry after each PR is merged.

## Image size

Currently, the resulting nginx image is 91.3MB and the upstream version is 142MB (35% reduction). In the future we could decrease it even more by using a bare base instead of Ubuntu 22.04.

### CODE REVIEW

File: rockcraft.yaml [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    plugin: autotools
+    source: https://nginx.org/download/nginx-1.23.3.tar.gz
+    build-packages:
+      - libgd-dev
+      - libpcre3-dev
+      - libssl-dev
+      - zlib1g-dev
+    stage-packages:
+      - libpcre3-dev
+      - libssl-dev
```
Comment (ghislainbourgeois) [inline rockcraft.yaml:24]: I don't think we want to stage the `dev` packages.
```suggestion
      - libpcre3
      - libssl3
```

**Review by ghislainbourgeois — CHANGES_REQUESTED** [review]

File: rockcraft.yaml (hunk 1 above)
Comment (gruyaume) [reply rockcraft.yaml]: Good catch, actually I think we don't even need them at all, `ca_certificates` should be sufficient. Fixed.

**Review by ghislainbourgeois — APPROVED** [review]

File: tests/integration/test_integration.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+class TestNginxRock(unittest.TestCase):
+    """Integration tests for the nginx rock."""
+
+    def test_given_nginx_container_is_running_when_http_get_then_nginx_welcome_message_is_returned(
+        self
+    ):
+        response = requests.get(f"{NGINX_DOCKER_URL}:{NGINX_DOCKER_PORT}")
+
+        assert response.status_code == 200
+        assert "Welcome to nginx!" in response.text
```
Comment (ghislainbourgeois) [inline tests/integration/test_integration.py:27]: Nice!

**Review by ghislainbourgeois — APPROVED** [review]

File: .github/workflows/publish.yaml [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+          name: nginx-rock
+
+      - name: Import and push to github package
+        run: |
+          version="$(cat rockcraft.yaml | yq e '.version')"
+          sudo skopeo \
+            --insecure-policy \
+            copy \
+            oci-archive:"${{ env.IMAGE_NAME }}_${version}_amd64.rock" \
+            docker-daemon:"${{ env.REGISTRY }}/canonical/${{ env.IMAGE_NAME }}:${version}"
```
Comment (cjdcordeiro) [inline .github/workflows/publish.yaml:44]:
```suggestion
            docker://${{ env.REGISTRY }}/canonical/${{ env.IMAGE_NAME }}:${version}
```

File: .github/workflows/publish.yaml [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+          docker push ${{ env.REGISTRY }}/canonical/${{ env.IMAGE_NAME }}:${version}
```
Comment (cjdcordeiro) [inline .github/workflows/publish.yaml:45]:
```suggestion
```

**Review by cjdcordeiro — APPROVED** [review]
nice. it lgtm. just left a small nitpick

**Pull request merged**